### ISSUE TITLE
How to train Flux Lora on multiple GPUs?

### ISSUE THREAD

**Issue by IAn2018cs** [ostris/ai-toolkit#73]
I need help training Flux Lora on multiple GPUs. The memory on a single GPU is not sufficient, so I want to train on multiple GPUs. However, configuring device: cuda:0,1 in the config file doesn't seem to work.

Could you please provide guidance on how to properly set up and run Flux Lora training across multiple GPUs? The current single-GPU memory limitation is preventing me from training effectively.

Any assistance or examples of multi-GPU configurations for Flux Lora would be greatly appreciated. Thank you!

**Comment by asizk** [comment]
I have the same problem. Have you solved it?

**Comment by WarAnakin** [comment]
i'm currently making changes to the scripts on my end to run mult-gpus. I have quite a bit of requests and 1 gpu doesn't cut it.
I know that the kohya version of flux can run multi-gpus

**Comment by cuba6112** [comment]
Also looking into this!

**Comment by Eng-ZeyadTarek** [comment]
Same problem here.

**Comment by skein12** [comment]
yeah same issue, testing on one GPU is working great but can't see myself using this in the future without multi GPU

**Comment by WarAnakin** [comment]
> One way I see to train on multiple GPUs at once is to create several .yaml files, each with a different GPU and a different part of the dataset. This would require splitting the dataset into multiple parts and then, after training, combining the resulting .safetensors weights into a single file. I wouldn’t know how to do that merge.
> 
> However, the ideal solution would be to modify the code so that it uses multiple GPUs with a single .yaml file.

this has already been done with some different scripts, all in all the functionality is there and accelerate can be setup for multi-gpu from the start. It's just a matter of enabling more processes, equivalent to the number of gpus and loading each one with the dataset, spread the batch size across all the gpus (this would make a batch size per device and total batch), all this needs to be done on the same machine id aka rank0

**Comment by Teapack1** [comment]
Hello, how does look with this issue ? Is there a way to train with 2 x T4 GPU on Kaggle ? thank you

**Comment by WarAnakin** [comment]
> Hello, is there a way to use multiple GPUs in the ai-toolkit config ? I'm trying to train with 2 x T4 GPU on Kaggle. thank you

not yet

**Comment by dydxdt** [comment]
Also looking forward to the multi-gpu solutions!

**Comment by zhini-web** [comment]
Also looking forward to the multi-gpu solutions!

**Comment by sushmitxo** [comment]
Yep pls implement multi gpu use

**Comment by WarAnakin** [comment]
> jwadow

That's not how you run multi-gpu training. Simply editing the config file won't work.

**Comment by vinch00** [comment]
+1

**Comment by sherlhw** [comment]
+1

**Comment by prp-e** [comment]
Isn't there any ways to use multiple GPU's yet?

**Comment by WarAnakin** [comment]
not yet

**Comment by firofame** [comment]
Error from kaggle

torch.OutOfMemoryError: CUDA out of memory. Tried to allocate 32.00 MiB. GPU 0 has a total capacity of 14.74 GiB of which 2.12 MiB is free. Process 3302 has 14.74 GiB memory in use. Of the allocated memory 14.33 GiB is allocated by PyTorch, and 287.48 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)

**Comment by WarAnakin** [comment]
> Error from kaggle
> 
> torch.OutOfMemoryError: CUDA out of memory. Tried to allocate 32.00 MiB. GPU 0 has a total capacity of 14.74 GiB of which 2.12 MiB is free. Process 3302 has 14.74 GiB memory in use. Of the allocated memory 14.33 GiB is allocated by PyTorch, and 287.48 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation. See documentation for Memory Management (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)

Hello,

You might want to reduce the batch size, or resolution.

**Comment by prp-e** [comment]
Sorry to ask again, is there any update on multi-gpu usage?

**Comment by WarAnakin** [comment]
i don't mean to upset any1 but if you are really in a rush to train something and you need to use multi-gpus, kohya supports it

**Comment by chenhh17** [comment]
I have made a progress to set the ('split_model_over_gpus', True)  in the ('model', OrderedDict([
               
` ('model', OrderedDict([
                    # huggingface model name or path
                    ('name_or_path', 'black-forest-labs/FLUX.1-schnell'),
                    ('assisstant_lora_path', 'ostris/FLUX.1-schnell-training-adapter'),
                    ('is_flux', True),
                    ('quantize', True),  # run 8bit mixed precision
                    ('split_model_over_gpus', True) # split the model over multiple gpus, Added by Alex so that the model can be run on multiple GPUs
                    #('low_vram', True),  # uncomment this if the GPU is connected to your monitors. It will use less vram to quantize, but is slower.
                ])),`
It can now properly split SD data on two GPU, but I quickly run into problems that:
`Error in LoRAModule lora_down
Expected all tensors to be on the same device, but found at least two devices, cuda:1 and cuda:0! (when checking argument for argument mat2 in method wrapper_CUDA_mm)`

Have anyone found a solution to this?

**Comment by spuro** [comment]
+1

**Comment by jwadow** [comment]
> Still nothing? Why is MultiGPU ignored in so many of these projects?

Most likely, the impossibility of simple writing and debugging of code due to the lack of multiple gpu on devs' pc, as well as changes in design features and hundreds of long tests.
And also the mass lack of multi-GPU among users like you and me.

**Comment by Oruli** [comment]
> > > Still nothing? Why is MultiGPU ignored in so many of these projects?
> > 
> > 
> > Most likely, the impossibility of simple writing and debugging of code due to the lack of multiple gpu on devs' pc, as well as changes in design features and hundreds of long tests. And also the mass lack of multi-GPU among users like you and me.
> 
> Have you guys tried diffusion-pipe, which works on multiple GPUs by splitting the model, exactly as expected

It doesn't work with 5090s. The dev told me it's related to a deepspeed issue, they are failing to fix it, waste of money buying them so far!

**Comment by Oruli** [comment]
This has been open for a year, is there actually anyone on the project who can at least acknowledge it and add to a todo or something?

**Comment by skein12** [comment]
just use bmaltais/kohya_ss this is never getting added lol

**Comment by an80sPWNstar** [comment]
any update on this? I love the simplicity of this lora trainer :)

**Comment by jordiferrero** [comment]
+1

**Comment by Oruli** [comment]
@jaretburkett I see you have a branch created for multi-gpu. Diffusion-pipe implementation using deepspeed is very good, he utilised your automagic algorithm can't you reuse code from there to get this sorted?

So many people needs this, and it's preventing us from using this project at all, which is a shame as it's much more active than diffusion-pipe.

**Comment by Oksitaine** [comment]
New about multi-gpu training in ostris ??
